src/mono/mono/mini/method-to-ir.c

```diff
@@ -7514,6 +7514,89 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 			if (cfg->gsharedvt_min && mini_is_gsharedvt_variable_signature (fsig))
 				GSHAREDVT_FAILURE (il_op);
 
+#ifdef MONO_ARCH_HAVE_SWIFTCALL
+			/*
+			 * We need to modify the signature of the swiftcall calli to account for the lowering of Swift structs.
+ 			 * This is done by replacing struct arguments on stack with a lowered sequence and updating the signature.
+			 */
+			if (fsig->pinvoke && mono_method_signature_has_ext_callconv (fsig, MONO_EXT_CALLCONV_SWIFTCALL)) {
+				g_assert (!fsig->hasthis); // Swift P/Invoke calls shouldn't contain 'this'
+				
+				n = fsig->param_count;
+				sp -= n;
+				// Save the old arguments				
+				MonoInst **old_args = g_newa (MonoInst*, n);
+				for (int idx_param = 0; idx_param < n; ++idx_param) {
+					old_args [idx_param] = sp [idx_param];
+				}
+
+				GArray *new_params = g_array_new (FALSE, FALSE, sizeof (MonoType*));
+				uint32_t new_param_count = 0;
+				MonoClass *swift_self = mono_class_try_get_swift_self_class ();
+				MonoClass *swift_error = mono_class_try_get_swift_error_class ();
+				/*
+				 * Go through the lowered arguments, if the argument is a struct, 
+				 * we need to replace it with a sequence of lowered arguments.
+				 * Also record the updated parameters for the new signature.
+				 */
+				for (int idx_param = 0; idx_param < n; ++idx_param) {
+					MonoType *ptype = fsig->params [idx_param];
+					MonoClass *klass = mono_class_from_mono_type_internal (ptype);
+
+					// SwiftSelf and SwiftError are special cases where we need to preserve the class information for the codegen to handle them correctly.
+					if (mono_type_is_struct (ptype) && !(klass == swift_self || klass == swift_error)) {
+						SwiftPhysicalLowering lowered_swift_struct = mono_marshal_get_swift_physical_lowering (ptype, FALSE);
+						if (!lowered_swift_struct.by_reference) {
+							// Create a new local variable to store the base address of the struct
+							MonoInst *struct_base_address =  mono_compile_create_var (cfg, mono_get_int_type (), OP_LOCAL);
+							CHECK_ARG (idx_param);
+							NEW_ARGLOADA (cfg, struct_base_address, idx_param);
+							MONO_ADD_INS (cfg->cbb, struct_base_address);
```

Comment:
Will this be optimized in the final codegen?

Reply:
How optimized would you expect the final codegen to be? The resulting assembly looks somewhat like this:
```
...
000000000000007c        ldr     d0, [x29, #0x10] // Load first lowered element
0000000000000080        ldr     x0, [x29, #0x18] // Load second lowered element
0000000000000084        blr     x1 // Call to native Swift func
...
```
where `x29` contains the base address of the struct.

Reply:
As a new local variable is introduced here, and I assume that `x_arg`, where `_arg` is the arg index, is moved to `x29`. I wonder if the final codegen contains the `mov` ins or utilizes the `x_arg` directly.

Reply:
Based on the codegen, I think that it just uses an offset from the beginning of stack pointer (`x29`). It doesn't store the new local variable in registers.